openmdao/components/subproblem_comp.py

```diff
@@ -0,0 +1,284 @@
+                             desc='Subproblem Component outputs')
+
+        # set other variables necessary for subproblem
+        self._prev_complex_step = False
+
+        self.prob_args = {'driver': driver,
+                          'comm': comm,
+                          'name': name,
+                          'reports': reports}
+
+        self.prob_args.update(prob_options)
+
+        self.model = model
+        self.list_inputs = inputs
+        self.list_outputs = outputs
+
+    def setup(self):
+        """
+        Perform some final setup and checks.
+        """
+        p = self._subprob = om.Problem(**self.prob_args)
+        p.model.add_subsystem('subsys', self.model, promotes=['*'])
+
+        p.setup(force_alloc_complex=self._problem_meta['force_alloc_complex'])
+        p.final_setup()
+
+        model_inputs = p.model.list_inputs(out_stream=None, prom_name=True,
+                                           units=True, shape=True, desc=True)
+        model_outputs = p.model.list_outputs(out_stream=None, prom_name=True,
+                                             units=True, shape=True, desc=True)
+
+        # store model inputs/outputs as dictionary with keys as the promoted name
+        model_inputs = {meta['prom_name']: meta for _, meta in model_inputs}
+        model_outputs = {meta['prom_name']: meta for _, meta in model_outputs}
+
+        self.options.update(_get_model_vars('inputs', self.list_inputs, model_inputs))
+        self.options.update(_get_model_vars('outputs', self.list_outputs, model_outputs))
+
+        inputs = self.options['inputs']
+        outputs = self.options['outputs']
+
+        # instantiate input/output name list for use in compute and
+        # compute partials
+        self._input_names = []
+        self._output_names = []
+
+        # remove the `prom_name` from the metadata and then store it for each
+        # input and output
+        for var, meta in inputs.items():
+            prom_name = meta.pop('prom_name')
+            self.add_input(var, **meta)
+            meta['prom_name'] = prom_name
+            self._input_names.append(var)
+
+        for var, meta in outputs.items():
+            prom_name = meta.pop('prom_name')
+            self.add_output(var, **meta)
+            meta['prom_name'] = prom_name
+            self._output_names.append(var)
+
+            for ip in self._input_names:
+                self.declare_partials(of=var, wrt=ip)
+
+    def compute(self, inputs, outputs):
+        """
+        Perform the subproblem system computation at run time.
+
+        Parameters
+        ----------
+        inputs : Vector
+            Unscaled, dimensional input variables read via inputs[key].
+        outputs : Vector
+            Unscaled, dimensional output variables read via outputs[key].
+        """
+        p = self._subprob
+
+        # switch subproblem to use complex IO if in complex step mode
+        if self.under_complex_step != self._prev_complex_step:
```

Comment:
This comment was marked as resolved with no explanation.  In general it's helpful to respond to all review comments in some way, especially if you disagree with a recommendation.  That way the reviewer at least knows why you didn't make a recommended change.

Reply:
I included the function you gave above, but I still thought it was necessary to keep the `if self.under_complex_step != self._prev_complex_step` conditional block because that way there won't be as many calls to `_set_complex_step_mode` whereas if you replace that block with `self._set_complex_step_mode(self._under_complex_step)` then it will be calling it every time `compute` is called.

I guess I thought that would cause it to perform slower, but I could be wrong. Do you think that would be possible?

Also, I wasn't aware of the responding to reviews rule of thumb. I'll be sure to do that in the future.

Reply:
I think that if you did the override of the _set_complex_step_mode method then you wouldn't have to do the check at all in compute, because the component will be in whatever mode is appropriate.  Basically overriding that method would just keep the subproblem in sync with whatever mode the component is in based on whatever mode is currently active.